### PR TITLE
ENH ModulusStable + fix_length

### PR DESCRIPTION
**Pull request by xir4n**
Adapted from Kymatio, librosa

### CODE REVIEW

File: murenn/dtcwt/nn.py [hunk 1]
```diff
@@ -38,38 +43,64 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
                 groups=in_channels,
                 padding="same",
             )
-            torch.nn.init.normal_(conv1d_j.weight)
+            torch.nn.init.normal_(conv1d_j.weight, std=1/math.sqrt(T))
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:46]: what's the rationale for this `1/sqrt(T)`? i'm guessing it has to do with expected value of the random energy response

are you testing it somewhere?

File: murenn/dtcwt/nn.py [hunk 2]
```diff
@@ -20,13 +23,15 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
         self.dtcwt = murenn.DTCWT(
             J=J,
             padding_mode=padding_mode,
+            normalize=True,
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:26]: isn't `normalize=True` already the default? are you saying there should be no default?

File: murenn/dtcwt/nn.py [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
+
+from .utils import fix_length
 
 
 class MuReNNDirect(torch.nn.Module):
     """
     Args:
-        J (int): Number of levels of DTCWT decomposition.
-        Q (int): Number of Conv1D filters at each level.
+        J (int): Number of levels (octaves) of the DTCWT decomposition.
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:11]: ideally, "in the DTCWT decomposition" (not "of")

File: murenn/dtcwt/nn.py [hunk 4]
```diff
@@ -38,38 +43,64 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
...
-            Wx_i = self.conv1d[j](bps[j].imag)
-            Ux = Wx_r ** 2 + Wx_i ** 2
-            y_j, _ = self.down[j](Ux)
+            Wx_j_r = self.conv1d[j](bps[j].real)
+            Wx_j_i = self.conv1d[j](bps[j].imag)
+            UWx_j = ModulusStable.apply(Wx_j_r, Wx_j_i)
+            UWx_j, _ = self.down[j](UWx_j)
 
-            B, _, N = y_j.shape
+            B, C, N = UWx_j.shape
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:72]: C is unused ?

File: murenn/dtcwt/nn.py [hunk 5]
```diff
@@ -38,38 +43,64 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
...
+
+class ModulusStable(torch.autograd.Function):
+    """Stable complex modulus
+
+    This class implements a modulus transform for complex numbers which is
+    stable with respect to very small inputs (z close to 0), avoiding
+    returning nans in all cases.
+
+    -------
+    Adapted from Kymatio: https://github.com/kymatio/kymatio/blob/main/kymatio/backend/torch_backend.py
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:88]: i don't recommend putting a URL here. "adapted from Kymatio" suffices
no need for a copyright notice, Kymatio and MuReNN are siblings 😊

File: murenn/dtcwt/nn.py [hunk 6]
```diff
@@ -38,38 +43,64 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
...
+
+        return torch.stack(output, dim=3)
+
+
+class ModulusStable(torch.autograd.Function):
+    """Stable complex modulus
+
+    This class implements a modulus transform for complex numbers which is
+    stable with respect to very small inputs (z close to 0), avoiding
+    returning nans in all cases.
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:85]: NaN's

File: tests/test_nn.py [hunk 7]
```diff
@@ -53,4 +60,30 @@ def test_multi_layers(J, Q, T, N):
             T=T,
             in_channels=x.shape[1],
         )
-        x = layer_i(x)
\ No newline at end of file
+        x = layer_i(x)
+
```
Comment (lostanlen) [inline tests/test_nn.py:64]: i'd love to see a simple test for `fix_length`

File: tests/test_nn.py [hunk 8]
```diff
@@ -38,7 +45,7 @@ def test_direct_diff():
         assert conv1d.weight.grad != None
 
 
-@pytest.mark.parametrize("J", list(range(1, 3)))
+@pytest.mark.parametrize("J", list(range(2, 3)))
```
Comment (lostanlen) [inline tests/test_nn.py:48]: or just `[2]` instead of `list(range(2, 3))` ?

**Review by lostanlen — CHANGES_REQUESTED** [review]

File: murenn/dtcwt/nn.py (hunk 1 above)
Comment (xir4n) [reply murenn/dtcwt/nn.py:46]: no, I didn't do any tests, this was a mistake.

**Comment by lostanlen** [comment]
@xir4n don't forget to re-request review whenever you're ready

**Comment by xir4n** [comment]
I'm unsure whether ModuluStable is a suitable activation function for a DNN...
Although it works for the scattering transform

**Comment by lostanlen** [comment]
i just know Zeghidour uses it!
https://arxiv.org/abs/1711.01161

and Fergal Cotter of course
https://arxiv.org/abs/1903.03137

and Muawiz on Parametric Scattering Networks
https://arxiv.org/abs/2107.09539

**Review by lostanlen — APPROVED** [review]
